text/0000-ques-in-main.md

```diff
@@ -0,0 +1,514 @@
+   the ability to pass an "exit status" up to some outer environment,
+   conventions for what that status means, and an expectation that a
+   diagnostic message will be generated when a program fails
+   due to a system error.
+1. We should avoid making life more complicated for people who don't
+   care; on the other hand, if the Easiest Thing is also the Right
+   Thing according to the platform convention, that is better all
+   around.
+
+Goal 1 dictates that the new return type for `main` will be
+`Result<T,E>` for some T and E.  To minimize the necessary changes to
+existing code that wants to start using `?` in `main`, T should be
+_allowed_ to be `()`, but other types in that position may also make
+sense.  The boilerplate shown [above][motivation] will work for any E
+that is `Display`, but that is probably too general for the stdlib; we
+propose only `Error` types should work by default.
+
+[grep]: http://man7.org/linux/man-pages/man1/grep.1.html
+
+## The `Termination` trait
+[the-termination-trait]: #the-termination-trait
+
+When `main` returns a nontrivial value, the runtime needs to know two
+things about it: what error message, if any, to print, and what value
+to pass to `std::process::exit`.  These are naturally encapsulated in
+a trait, which we are tentatively calling `Termination`, with this
+signature:
+
+``` rust
+trait Termination {
+    fn write_diagnostic(&self, progname: &str, stream: &mut Write) -> ();
+    fn exit_status(&self) -> i32;
+}
+```
+
+`write_diagnostic` shall write a complete diagnostic message for
+`self` to its `stream` argument.  If it produces no output, there will
+be no output.  `stream` will be the same stream that `panic` messages
+from the main thread would go to, which is normally "standard error".
+(An application-controllable override mechanism may make sense,
+[see below][squelching-diagnostics].)
+
+The `progname` argument is a short version of the program's name
+(abstractly, what you would type at the command line to start the
+program, assuming it were in `PATH`; concretely, the basename of
+`argv[0]`, with any trailing `.exe` or `.com` chopped off on Windows,
+but not on other platforms).  This makes it easy to generate
+diagnostics in the style conventional for "Unixy" systems, e.g.
+
+    grep: /etc/shadow: Permission denied
+
+`exit_status` shall convert `self` to a platform-specific exit code,
+conveying at least a notion of success or failure.  The return type is
+`i32` to match [std::process::exit][] (which probably calls the C
+library's `exit` primitive), but (as already documented for
+`process::exit`) on "most Unix-like" operating systems, only the low 8
+bits of this value are significant.
+
+[std::process::exit]: https://doc.rust-lang.org/std/process/fn.exit.html
+
+## Changes to `main`
+[changes-to-main]: #changes-to-main
+
+From the perspective of the code that calls `main`, its signature is
+now generic:
+
+``` rust
+fn<T: Termination> main() -> T { ... }
```

Comment:
I think this is pseudosyntax? Since the only way this proposal can work is `fn main() -> T` where `T: Termination`, but `T` has to be a concrete type.